lib/cached_network_image.dart

```diff
@@ -215,57 +315,94 @@ class CacheObject {
     return null;
   }
 
+  DateTime touched;
+  String url;
+
   Object lock;
   Map _map;
 
-  CacheObject() {
+  CacheObject(String url) {
+    this.url = url;
     _map = new Map();
+    touch();
     lock = new Object();
   }
 
-  CacheObject.fromMap(Map map) {
+  CacheObject.fromMap(String url, Map map) {
+    this.url = url;
     _map = map;
+
+    if (_map.containsKey("touched")) {
+      touched = new DateTime.fromMillisecondsSinceEpoch(_map["touched"]);
+    } else {
+      touch();
+    }
+
     lock = new Object();
   }
 
   Map toMap() {
     return _map;
   }
 
+  touch() {
+    touched = new DateTime.now();
+    _map["touched"] = touched.millisecondsSinceEpoch;
+  }
+
   setDataFromHeaders(Map<String, String> headers) async {
+    //Without a cache-control header we keep the file for a week
+    var ageDuration = new Duration(days: 7);
+
     if (headers.containsKey("cache-control")) {
       var cacheControl = headers["cache-control"];
       var controlSettings = cacheControl.split(", ");
       controlSettings.forEach((setting) {
         if (setting.startsWith("max-age=")) {
           var validSeconds =
-          int.parse(setting.split("=")[1], onError: (source) => 0);
+              int.parse(setting.split("=")[1], onError: (source) => 0);
           if (validSeconds > 0) {
-            _map["validTill"] = new DateTime.now()
-                .add(new Duration(seconds: validSeconds))
-                .millisecondsSinceEpoch;
+            ageDuration = new Duration(seconds: validSeconds);
           }
         }
       });
     }
 
+    _map["validTill"] =
+        new DateTime.now().add(ageDuration).millisecondsSinceEpoch;
+
     if (headers.containsKey("etag")) {
       _map["ETag"] = headers["etag"];
     }
 
+    var fileExtension = "";
     if (headers.containsKey("content-type")) {
       var type = headers["content-type"].split("/");
-      if (type[0] == "image") {
-        if (filePath == null || !filePath.endsWith(type[1])) {
-          Directory directory = await getTemporaryDirectory();
-          var folder = new Directory("${directory.path}/imagecache");
-          if (!(await folder.exists())) {
-            folder.createSync();
-          }
-          var fileName = "${new Uuid().v1()}.${type[1]}";
-          _map["path"] = "${folder.path}/${fileName}";
-        }
+      if (type.length == 2) {
+        fileExtension = ".${type[1]}";
+      }
+    }
+
+    if(filePath != null && !filePath.endsWith(fileExtension)){
+      removeOldFile(filePath);
+      _map["path"] = null;
+    }
+
+    if(filePath == null){
+      Directory directory = await getTemporaryDirectory();
+      var folder = new Directory("${directory.path}/cache");
+      if (!(await folder.exists())) {
+      folder.createSync();
```

Comment:
another place where we create folders ...?

Reply:
Same folder as earlier. It is probably better to only do it in one place. I did this because otherwise it broke when the file was gone but the filepath was already determined in the cache manager.